cmd/syncthing/gui.go

```diff
@@ -314,49 +351,22 @@ func (s *apiService) Serve() {
 
 	guiCfg := s.cfg.GUI()
 
-	// Wrap everything in CSRF protection. The /rest prefix should be
-	// protected, other requests will grant cookies.
-	handler := csrfMiddleware(s.id.String()[:5], "/rest", guiCfg, mux)
-
-	// Add our version and ID as a header to responses
-	handler = withDetailsMiddleware(s.id, handler)
-
-	// Wrap everything in basic auth, if user/password is set.
-	if len(guiCfg.User) > 0 && len(guiCfg.Password) > 0 {
-		handler = basicAuthAndSessionMiddleware("sessionid-"+s.id.String()[:5], guiCfg, handler)
-	}
-
-	// Redirect to HTTPS if we are supposed to
-	if guiCfg.UseTLS() {
-		handler = redirectToHTTPSMiddleware(handler)
-	}
-
-	// Add the CORS handling
-	handler = corsMiddleware(handler, guiCfg.InsecureAllowFrameLoading)
-
-	if addressIsLocalhost(guiCfg.Address()) && !guiCfg.InsecureSkipHostCheck {
-		// Verify source host
-		handler = localhostMiddleware(handler)
-	}
-
-	handler = debugMiddleware(handler)
-
-	srv := http.Server{
-		Handler: handler,
-		// ReadTimeout must be longer than SyncthingController $scope.refresh
-		// interval to avoid HTTP keepalive/GUI refresh race.
-		ReadTimeout: 15 * time.Second,
+	handlers := make([]http.Handler, 0, len(listeners))
+	for i, guiListener := range guiCfg.GUIListeners() {
+		handlers = append(handlers, s.getHandler(mux, &guiListener, listeners[i]))
 	}
 
 	s.fss = newFolderSummaryService(s.cfg, s.model)
 	defer s.fss.Stop()
 	s.fss.ServeBackground()
 
-	l.Infoln("GUI and API listening on", listener.Addr())
-	l.Infoln("Access the GUI via the following URL:", guiCfg.URL())
+	for _, guiListener := range guiCfg.GUIListeners() {
+		l.Infoln("GUI and API listening on", guiListener.Address)
+		l.Infoln("Access the GUI via the following URL:", guiCfg.URLFromGUIListener(guiListener))
+	}
```

Comment:
So this is ok for this PR I guess, but in general, I'd expect the Access the GUI to be printed only once for something that is not a unix socket. I know non-unix socket is not possible now, but just thinking ahead.

Reply:
So, for this PR, I propose to only add a couple of comments to keep in mind that further work will need to be done here when we add UNIX socket support.

Also, to be sure : if one user configures several TCP listeners, do you definitely expect to print the Access URL only once for the first TCP listener ?

Reply:
Perhaps you are right, it's most likely fine to print them all. We already print a ton of stuff, so this won't make it worse.